### PR TITLE
Reuse collections in DfaMatcherBuilder

### PR DESCRIPTION
**Pull request by JamesNK**
Before:
```
 Method |     Mean |    Error |   StdDev |  Op/s |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
------- |---------:|---------:|---------:|------:|---------:|---------:|---------:|----------:|
    Dfa | 357.2 ms | 7.020 ms | 6.567 ms | 2.799 | 937.5000 | 875.0000 | 812.5000 | 195.18 MB |
```

After:
```
 Method |     Mean |    Error |   StdDev |  Op/s |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
------- |---------:|---------:|---------:|------:|---------:|---------:|---------:|----------:|
    Dfa | 308.8 ms | 5.961 ms | 5.284 ms | 3.238 | 937.5000 | 937.5000 | 812.5000 | 150.32 MB |
```

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs [hunk 1]
```diff
@@ -9,6 +9,33 @@
 
 namespace Microsoft.AspNetCore.Routing.Matching
 {
+    internal ref struct CandidateBuilderState
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs]: I started doing this thing to see what it felt like and quickly felt it was stupid. Validate this is stupid for me and I'll move this state to fields on DfaMatcherBuilder

File: src/Microsoft.AspNetCore.Routing/Matching/HttpMethodMatcherPolicy.cs [hunk 2]
```diff
@@ -362,7 +362,7 @@ public override bool Equals(object obj)
             public override int GetHashCode()
             {
                 var hash = new HashCodeCombiner();
-                hash.Add(IsCorsPreflightRequest);
+                hash.Add(IsCorsPreflightRequest ? 1 : 0);
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matching/HttpMethodMatcherPolicy.cs]: There was no bool overload so it was boxing to object.

**Comment by JamesNK** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs [hunk 3]
```diff
@@ -370,20 +414,20 @@ internal Candidate[] CreateCandidates(IReadOnlyList<Endpoint> endpoints)
         // internal for tests
         internal Candidate CreateCandidate(Endpoint endpoint, int score)
         {
-            var assignments = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-            var slots = new List<KeyValuePair<string, object>>();
-            var captures = new List<(string parameterName, int segmentIndex, int slotIndex)>();
-            (string parameterName, int segmentIndex, int slotIndex) catchAll = default;
+            _assignments.Clear();
```
Comment (drieseng) [inline src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs]: Do you want to clear these lists at all when **endpoint** is not a **RouteEndpoint**?
In that case you could create a **Candidate** instance with only **Array.Empty&lt;T&gt;** instances
No idea how common this is though.

**Pull request merged**